### PR TITLE
Reset state

### PR DESCRIPTION
**Pull request by gregw**
The complex conditional behaviour of the closed state indicates that it is really two states: reset and closed.  The reset state is entered when a reset is sent and then transitions to the closed state only when the peer acknowledges the stream is closed with either an END_STREAM flag or RST_STREAM frame.

By making RST_STREAM an acknowledged frame there is no need for indeterminate behaviour with timeouts to handle frames sent while a RST_STREAM is in transit.

### CODE REVIEW

File: draft-ietf-httpbis-http2.xml [hunk 1]
```diff
@@ -773,34 +773,34 @@ HTTP2-Settings    = token68
...
-     |         |             +--------+             |         |
-     |         |     recv ES |        | send ES     |         |
-     |  send H |     ,-------|  open  |-------.     | recv H  |
-     |         |    /        |        |        \    |         |
-     |         v   v         +--------+         v   v         |
-     |     +----------+          |           +----------+     |
-     |     |   half   |          |           |   half   |     |
-     |     |  closed  |          | send R/   |  closed  |     |
-     |     | (remote) |          | recv R    | (local)  |     |
+     ,-----| reserved |          | semd H/   | reserved |-----.
```
Comment (Scottmitch) [inline draft-ietf-httpbis-http2.xml]: Semd -> send

File: draft-ietf-httpbis-http2.xml [hunk 2]
```diff
@@ -995,64 +995,64 @@ HTTP2-Settings    = token68
                 target="FlowControl">stream level flow control limits</xref>.
               </t>
               <t>
-                A stream can transition from this state to "closed" by sending a frame that contains
-                an END_STREAM flag, or when either peer sends a <x:ref>RST_STREAM</x:ref> frame.
+                A stream transitions from this state to "closed" when it receives  
+		an END_STREAM flag or a <x:ref>RST_STREAM</x:ref> frame. A stream transitions from this
```
Comment (Scottmitch) [inline draft-ietf-httpbis-http2.xml]: Indentation is a bit off

File: draft-ietf-httpbis-http2.xml [hunk 3]
```diff
@@ -995,64 +995,64 @@ HTTP2-Settings    = token68
...
-                If this state is reached as a result of sending a <x:ref>RST_STREAM</x:ref> frame,
-                the peer that receives the <x:ref>RST_STREAM</x:ref> might have already sent - or
-                enqueued for sending - frames on the stream that cannot be withdrawn.  An endpoint
-                MUST ignore frames that it receives on closed streams after it has sent a
-                <x:ref>RST_STREAM</x:ref> frame.  An endpoint MAY choose to limit the period over
-                which it ignores frames and treat frames that arrive after this time as being in
-                error.
+                Flow controlled frames (i.e., <x:ref>DATA</x:ref>) received in this state
+                are counted toward the connection flow control window, even though their content
+		is ignored as the sender will consider the frames to count
```
Comment (Scottmitch) [inline draft-ietf-httpbis-http2.xml]: Indentation

File: draft-ietf-httpbis-http2.xml [hunk 4]
```diff
@@ -995,64 +995,64 @@ HTTP2-Settings    = token68
...
+              </t>
+              <t>
+                An endpoint MUST NOT send frames other than <x:ref>PRIORITY</x:ref> on a closed
+                stream.  An endpoint that receives any frame other than <x:ref>PRIORITY</x:ref>
+                or <x:ref>RST_STREAM</x:ref> in this state MUST treat that as a 
+		<xref target="StreamErrorHandler">stream error</xref> of type
+                <x:ref>STREAM_CLOSED</x:ref>.  
+              </t>
+              <t>
+                <x:ref>PRIORITY</x:ref> frames can be sent on closed streams to prioritize streams
```
Comment (Scottmitch) [inline draft-ietf-httpbis-http2.xml]: Should `closed` be in quotes for consistency when referring to states?

Comment (gregw) [inline draft-ietf-httpbis-http2.xml]: Probably, but I noticed that it is other states are not consistenty quoted, so was following the practise of what was before me.... but will change to consistency as I think that is best.

Comment (gregw) [inline draft-ietf-httpbis-http2.xml]: Looks like the convention is to use "closed" when talking about the state, but just closed when talking about the stream.   Document is full of closed streams so I wont change those.

thanks

**Comment by gregw** [comment]
I just noticed that my branch picked up a couple of reverted commits from previous PRs.   So IF this is accepted, I'll remake it on a fresh branch without the noise commits.  But I'll leave it as is for now to see if it has a hope first.

**Comment by gregw** [comment]
I couldn't stand the mess, so I have recreated without noise reverts and with Scotts feedback in https://github.com/http2/http2-spec/pull/701

**Pull request closed without merging**